dfusion_rust_core/src/models/order.rs

```diff
@@ -47,6 +47,90 @@ impl Order {
             sell_amount,
         }
     }
+
+    /// Calculates the utility of an order based on a buy and sell price as well
+    /// as the executed buy amount.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` utility result if the inputs were valid and it
+    /// it produced a valid positive utility.
+    pub fn utility(
+        &self,
+        buy_price: u128,
+        sell_price: u128,
+        exec_buy_amount: u128,
+    ) -> Option<U256> {
+        // first lets convert everything into U256s
+        let buy_price = u128_to_u256(buy_price);
+        let sell_price = u128_to_u256(sell_price);
+        let exec_buy_amount = u128_to_u256(exec_buy_amount);
+        let buy_amount = u128_to_u256(self.buy_amount);
+        let sell_amount = u128_to_u256(self.sell_amount);
+        let fee_denominator = U256::from(2);
+
+        // adj_sell_amount =
+        //   (
+        //     (
+        //       (exec_buy_amount * buy_price) / (fee_denominator - 1)
+        //     ) *
+        //     fee_denominator
+        //   ) /
+        //   sell_price
+        //
+        // utility =
+        //   (
+        //     (exec_buy_amount * buy_price - adj_sell_amount * buy_amount) *
+        //     buy_price
+        //   ) /
+        //   sell_amount
+        let adj_sell_amount = exec_buy_amount
+            .checked_mul(buy_price)?
+            .checked_div(fee_denominator - 1)?
+            .checked_mul(fee_denominator)?
+            .checked_div(sell_price)?;
+        let utility = exec_buy_amount
+            .checked_mul(buy_price)?
+            .checked_sub(adj_sell_amount.checked_mul(buy_amount)?)?
+            .checked_mul(buy_price)?
+            .checked_div(sell_amount)?;
+
+        Some(utility)
+    }
+
+    /// Calculates the disregarded utility of an order based on a buy and sell
+    /// price as well as the executed sell amount.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` disregarded utility result if the inputs were valid
+    /// and it it produced a valid positive utility.
+    pub fn disregarded_utility(
+        &self,
+        buy_price: u128,
+        sell_price: u128,
+        exec_sell_amount: u128,
+    ) -> Option<U256> {
+        // first lets convert everything into U256s
+        let buy_price = u128_to_u256(buy_price);
+        let sell_price = u128_to_u256(sell_price);
+        let exec_sell_amount = u128_to_u256(exec_sell_amount);
+        let buy_amount = u128_to_u256(self.buy_amount);
+        let sell_amount = u128_to_u256(self.sell_amount);
+
+        // disregarded_utility =
+        //   (
+        //     (sell_price * sell_amount - buy_price * buy_amount) *
+        //     (sell_amount - exec_sell_amount)
+        //   ) /
+        //   sell_amount
+
+        sell_price
+            .checked_mul(sell_amount)?
+            .checked_sub(buy_price.checked_mul(buy_amount)?)?
```

Comment:
ubernit: I would like to swap the order of `buy_amount` and `buy_price` here because, although it appears there is a symmetry in the naming convention, the term itself is a result of cross multiplication. I was hoping to preserve some of the intuition in the calculation without too many mathe-magical tricks (like multiplicative commutativity of most number systems)

Reply:
Easy fix